src/geeknn/ordination/_base.py

```diff
@@ -0,0 +1,282 @@
+
+    def properties_to_array(self, properties: list[str]) -> NDArray:
+        """Return specified properties as a 2-D numpy array."""
+        return np.array([f.get_properties(properties) for f in self.features])
+
+    @classmethod
+    def from_ee_feature_collection(
+        cls, fc: ee.FeatureCollection, num_threads: int = -1, chunk_size: int = 5000
+    ):
+        """Create a client-side FeatureCollection from an ee.FeatureCollection
+        using multiple threads and chunking to avoid memory issues.
+        """
+        size = fc.size().getInfo()
+        info = fc.limit(0).getInfo()
+        chunks = [fc.toList(chunk_size, i) for i in range(0, size, chunk_size)]
+
+        with Parallel(n_jobs=num_threads, backend="threading") as p:
+            chunk_data = p(delayed(chunk.getInfo)() for chunk in chunks)
+
+        info.update({"features": list(chain.from_iterable(chunk_data))})
+        return cls(**info)
+
+
+class Raw:
+    def __init__(self, k=1, max_duplicates=None):
+        self.k = k
+        self.max_duplicates = max_duplicates if max_duplicates is not None else 5
+
+    def transform_image(self, X_image: ee.Image) -> ee.Image:
+        """No-op placeholder."""
+        return X_image
+
+    def transform_fc(self, fc: ee.FeatureCollection) -> ee.FeatureCollection:
+        """No-op placeholder."""
+        return fc
+
+    @property
+    def k_nearest(self):
+        """Return the number of requested nearest neighbors plus max_duplicates
+        in order to account for non-independent samples.
+        """
+        return self.k + self.max_duplicates
+
+    def get_ids(self, fc: FeatureCollection, id_field: str) -> list[Any]:
+        """Return the IDs from the feature collection."""
+        return list(map(int, fc.aggregate_array(id_field)))
+
+    def _set_classifier(
+        self,
+        *,
+        fc: ee.FeatureCollection,
+        class_property: ee.String,
+        input_properties: ee.List,
+    ) -> None:
+        """Set the EE minimumDistance classifier for this estimator."""
+        self.clf = ee.Classifier.minimumDistance(
+            metric="euclidean",
+            kNearest=self.k_nearest,
+        ).train(
+            features=fc,
+            classProperty=class_property,
+            inputProperties=input_properties,
+        )
+
+    def train(
+        self, *, fc: ee.FeatureCollection, id_field: str, X_columns: list[str], **_
+    ):
+        """Train the estimator, storing the needed server-side objects used
+        in prediction.
+        """
+        self.id_field = ee.String(id_field)
+        self.X_columns = ee.List(X_columns)
+        self._set_classifier(
+            fc=fc, class_property=self.id_field, input_properties=self.X_columns
+        )
+        return self
+
+    def predict(self, X_image: ee.Image, mode: str = "CLASSIFICATION"):
+        """Predict the nearest neighbors for the given covariate (X) image."""
+        self.clf = self.clf.setOutputMode(ee.String(mode))
```

Comment:
Thanks. That one was a silly error ...

Reply:
Gotcha!

This may reveal my lack of understanding of how this all fits together, but I'm curious why there's no `mode` option for `predict_fc`? Does regression make sense in an image context but not in a feature context for some reason?

Reply:
> This may reveal my lack of understanding of how this all fits together

As always, your understanding is spot on.  I just haven't yet done anything with `REGRESSION` mode with feature collections.  For me, `REGRESSION` is a bit of a misnomer for `ee.Classifier.minimumDistance` as it's just returning the distances from each of the neighbors.  `sknnr` is way further ahead in this manner right now in that it (optionally) uses the NN distances in prediction of attributes.  For now, `geeknn` is mostly using just the neighbors in prediction and typically using equal weighting or our default k=7 weighting and those are separate modules, currently only in the JS version of this package.  For immediate needs, I just need to return neighbor IDs, but I think it makes sense to implement attribute prediction, at which point we will need to call `predict` in both `CLASSIFICATION` and `REGRESSION` modes to get the needed components for both images and feature collections.  I have a note to myself to create an issue to address this.